server/src/main/java/org/elasticsearch/index/codec/vectors/ES816BinaryQuantizedVectorsWriter.java

```diff
@@ -0,0 +1,1000 @@
+            correctionsBuffer.rewind();
+        }
+    }
+
+    private void writeSortingField(
+        FieldWriter fieldData,
+        float[] clusterCenter,
+        int maxDoc,
+        Sorter.DocMap sortMap,
+        BinaryQuantizer scalarQuantizer
+    ) throws IOException {
+        final int[] ordMap = new int[fieldData.getDocsWithFieldSet().cardinality()]; // new ord to old ord
+
+        DocsWithFieldSet newDocsWithField = new DocsWithFieldSet();
+        mapOldOrdToNewOrd(fieldData.getDocsWithFieldSet(), sortMap, null, ordMap, newDocsWithField);
+
+        // write vector values
+        long vectorDataOffset = binarizedVectorData.alignFilePointer(Float.BYTES);
+        writeSortedBinarizedVectors(fieldData, clusterCenter, ordMap, scalarQuantizer);
+        long quantizedVectorLength = binarizedVectorData.getFilePointer() - vectorDataOffset;
+
+        float centroidDp = VectorUtil.dotProduct(clusterCenter, clusterCenter);
+        writeMeta(fieldData.fieldInfo, maxDoc, vectorDataOffset, quantizedVectorLength, clusterCenter, centroidDp, newDocsWithField);
+    }
+
+    private void writeSortedBinarizedVectors(FieldWriter fieldData, float[] clusterCenter, int[] ordMap, BinaryQuantizer scalarQuantizer)
+        throws IOException {
+        byte[] vector = new byte[BQVectorUtils.discretize(fieldData.fieldInfo.getVectorDimension(), 64) / 8];
+        int correctionsCount = scalarQuantizer.getSimilarity() != EUCLIDEAN ? 3 : 2;
+        final ByteBuffer correctionsBuffer = ByteBuffer.allocate(Float.BYTES * correctionsCount).order(ByteOrder.LITTLE_ENDIAN);
+        for (int ordinal : ordMap) {
+            float[] v = fieldData.getVectors().get(ordinal);
+            float[] corrections = scalarQuantizer.quantizeForIndex(v, vector, clusterCenter);
+            binarizedVectorData.writeBytes(vector, vector.length);
+            for (int i = 0; i < corrections.length; i++) {
+                correctionsBuffer.putFloat(corrections[i]);
+            }
+            binarizedVectorData.writeBytes(correctionsBuffer.array(), correctionsBuffer.array().length);
+            correctionsBuffer.rewind();
+        }
+    }
+
+    private void writeMeta(
+        FieldInfo field,
+        int maxDoc,
+        long vectorDataOffset,
+        long vectorDataLength,
+        float[] clusterCenter,
+        float centroidDp,
+        DocsWithFieldSet docsWithField
+    ) throws IOException {
+        meta.writeInt(field.number);
+        meta.writeInt(field.getVectorEncoding().ordinal());
+        meta.writeInt(field.getVectorSimilarityFunction().ordinal());
+        meta.writeVInt(field.getVectorDimension());
+        meta.writeVLong(vectorDataOffset);
+        meta.writeVLong(vectorDataLength);
+        int count = docsWithField.cardinality();
+        meta.writeVInt(count);
+        if (count > 0) {
+            final ByteBuffer buffer = ByteBuffer.allocate(field.getVectorDimension() * Float.BYTES).order(ByteOrder.LITTLE_ENDIAN);
```

Comment:
The current format only allows a single clusterCenter.  If we want to add multiple clusters in the future, we need to modify the format.

Should we redesign the current format to allow multiple cluster centers?